[Swift] Add loops concept exercise

wneumann opened the pull request:
Added the loops concept exercise.

I have similar concerns with the story here as I have for tuples as they are both based on Santa Claus, so I worry about inclusivity with it.

stillleben commented on languages/swift/exercises/concept/loops/.docs/instructions.md:
````diff
@@ -24,29 +24,59 @@ hashIDs(ids)
 ```
 
-## 2. Compute the ranking level for a hashed ID
+## 2. Compute the digital sum of a number
````
Nice idea to make it a separate task.

stillleben commented on languages/swift/exercises/concept/loops/.docs/instructions.md:
```diff
@@ -24,29 +24,59 @@ hashIDs(ids)
+
+5217 has a digital sum of 5 + 2 + 1 + 7 = 15 but 5217 is not a multiple of 15, so the sequence is empty and the ranking level of 5217 is 0.
+
+8118 has a digital sum of 8 + 1 + 1 + 8, and 8118 / 18 = 451.  
```
```suggestion
8118 has a digital sum of 8 + 1 + 1 + 8 = 18, and 8118 / 18 = 451.  
```

stillleben commented on languages/swift/exercises/concept/loops/README.md:
```diff
@@ -24,27 +24,57 @@ hashIDs(ids)
+
+5217 has a digital sum of 5 + 2 + 1 + 7 = 15 but 5217 is not a multiple of 15, so the sequence is empty and the ranking level of 5217 is 0.
+
+8118 has a digital sum of 8 + 1 + 1 + 8, and 8118 / 18 = 451.  
```
```suggestion
8118 has a digital sum of 8 + 1 + 1 + 8 = 18, and 8118 / 18 = 451.  
```

wneumann commented:
I've simplified the ranking exercise and split it into two exercises, but I'm still not crazy about the way it reads. It looks a lot more complex than it actually is.

stillleben commented:
> I've simplified the ranking exercise and split it into two exercises, but I'm still not crazy about the way it reads. It looks a lot more complex than it actually is.

I don't see a problem there. The algorithm needs some upfront explanation and better to have it more verbose with the examples instead of having a short but unclear/ambiguous description, that raises more questions.

stillleben commented on languages/swift/exercises/concept/loops/Sources/loops/loops.swift:
```diff
@@ -0,0 +1,15 @@
+func hashIDs(_ ids: [Int]) -> [(Int, Int)] {
+  fatalError("hashIDs not implemented -- place your implementation here")
```
After looking into another PR from you I find the errorMessage `Please implement the ... function` more friendly and straight forward :)

ErikSchierboom commented on languages/swift/exercises/concept/loops/.docs/introduction.md:
```diff
@@ -0,0 +1,97 @@
+// 7 / 2 = 3
+// 11 / 2 = 5
+// Done with numbers
+
+
+for char in word {
+  if "aeiou".contains(char) {
+    print(char, terminator: "")
+  }
+}
+print(" - those are all the vowels")
+
+// prints:
+// ueaiaiiieiaioiou - those are all the vowels
+
```
Does this example introduce anything different that the student needs to know about to solve the exercise? If not, I would suggest to remove it.

wneumann replied:
The `for number in numbers` does, but I'll pull the chars part.

ErikSchierboom commented on languages/swift/exercises/concept/loops/.docs/introduction.md:
````diff
@@ -0,0 +1,97 @@
+
+```
+
+If one wants to execute a loop a specified number of times, a for-in loop can be used with a range supplied for the sequence to iterate over:
+
+```swift
+for i in 1...3 {
+  print("i: \(i)")
+}
+
+// prints:
+// i: 1
+// i: 2
+// i: 3
+```
````
Looking at the example implementation, it looks like this feature is not needed to solve the exercise. As such, it can be removed from this document.

wneumann replied:
Removed and saved for the after.

ErikSchierboom commented on languages/swift/exercises/concept/loops/.docs/introduction.md:
```diff
@@ -0,0 +1,97 @@
+
+## Control transfer
+
+Sometimes it is necessary to skip to the next iteration of a loop early, without completing the rest of the statements in the current iteration of the loop. The `continue` keyword can be used for this. When `continue` is executed, the loop jump to the next check to see if the next iteration of the loop can be run, i.e. the `while` in while and repeat-while loops or the check if there's another el ...
+
+Similarly, the `break` keyword can be used to exit a loop early, without executing the rest of the current iteration. When a `break` is executed, the loop is exited and execution picks up with the first line of code following the loop.
```
This introduces new syntax, which means that an example should be added.

wneumann replied:
Brought back in from the after.

ErikSchierboom commented on languages/swift/exercises/concept/loops/.docs/after.md:
````diff
@@ -0,0 +1,275 @@
+// 7
+```
+
+If one wants to execute a loop a specified number of times, a for-in loop can be used with a range supplied for the sequence to iterate over:
````
Are students familiar with ranges?

wneumann replied:
They will be. There is a ranges exercise scheduled to go between conditionals and loops, but it hasn't been written yet. I do have to add ranges to the prerequisites though.

ErikSchierboom replied:
👍

ErikSchierboom commented on languages/swift/exercises/concept/loops/.docs/instructions.md:
```diff
@@ -0,0 +1,85 @@
+- Every child in the world is represented in Santa's privacy-preserving database as an integer ID number in the range [1, 4,000,000,000).
+- The ranking level of each child can be encoded by using the magic hash to transform the ID into a hashed ID.
+- The ranking level can then be extracted from the hashed ID.
+
+## Tasks
```
```suggestion
```

ErikSchierboom commented on languages/swift/exercises/concept/loops/.docs/instructions.md:
```diff
@@ -0,0 +1,85 @@
+
+## 1. Hash the IDs from the database
+
+Your first task is to implement the function `hashIDs(\_:) which, given an array of ID numbers, takes each ID, transforms it with Santa's magic hash, and constructs a tuple of the original ID and the hashed ID, returning an array of these tuples.
```
```suggestion
Your first task is to implement the function `hashIDs(\_:)` which, given an array of ID numbers, takes each ID, transforms it with Santa's magic hash, and constructs a tuple of the original ID and the hashed ID, returning an array of these tuples.
```

ErikSchierboom commented on languages/swift/exercises/concept/loops/.docs/instructions.md:
```diff
@@ -0,0 +1,85 @@
+You are an elf working in Santa Claus' logistics division and you have been given a pair of tasks from the boss to help determine the gifts that Santa will deliver.
+
+It is believed that Santa sorts the various children of the world into naughty and nice to determine who gets presents, but that's not the case. What Santa really does is rank the children on a scale from 0 on up, where the higher the rank, the better the present the child gets. This ranking is based on an arcane formula that only Santa fully understands. But when it comes to the implementation o ...
```
@iHiD Thoughts about this story?

iHiD replied:
I like it 🙂

ErikSchierboom commented on languages/swift/exercises/concept/loops/.docs/instructions.md:
```diff
@@ -0,0 +1,85 @@
+
+## 3. Compute the ranking level for a hashed ID
+
+For this task, you will need to implement the function, `rankingLevel(_:)` that takes a hashed ID as input and computes the ranking level for the given input in terms of what is known as a Harshad number.
+
+A Harshad number is simply a number that is evenly divisible by the sum of its digits. So 1475 is not a Harshad number because 1 + 4 + 7 + 5 = 17, and 1475 is not divisible by 17. However, 8118 is a Harshad number because 8 + 1 + 1 + 8 = 18 and 8118 is divisible by 18.
+
+Given this definition of a Harshad number, the ranking level of a number is the length of the sequence of _distinct_ Harshad numbers where each element of the sequence is equal to the previous element divided by the sum of its digits.
+
+For example, consider the numbers 5217, 8118, and 7560:
+
+5217 has a digital sum of 5 + 2 + 1 + 7 = 15 but 5217 is not a multiple of 15, so the sequence is empty and the ranking level of 5217 is 0.
+
+8118 has a digital sum of 8 + 1 + 1 + 8 = 18, and 8118 / 18 = 451.  
+451 has a digital sum of 4 + 5 + 1 = 10, but 451 is not a multiple of 10, so we stop the process.
+
+Here, our sequence is just [8118]. Its length is 1 and so the ranking level is 1.
+
+7560 has a digital sum = 7 + 5 + 6 + 0 = 18, and 7560 / 18 => 420.  
+420 has a digital sum of 4 + 2 + 0 = 6 and 420 / 6 => 70.  
+70 has a digital sum of 7 + 0 = 7 and 70 / 7 => 10.  
+10 has a digital sum of 1 + 0 = 1 and 10 / 1 => 10, but this is not a distinct number, it is already in our sequence, so we stop the process.
+
+Here our sequence is [7560, 420, 70, 10]. It's length is 4 and thus our ranking level is 4.
```
On the one hand I really like this problem, but I feel like this type of problem (an algorithmic one) is perhaps better suitable for a practice exercise. While I feel it is fun to work on, it took be some time to fully understand what was being asked of me. In general, we want the concept exercises to be easy to understand

According to [the specs](https://github.com/exercism/v3/blob/master/docs/concept-exercises.md):

> A seasoned developer in Language X should be able to work through all the Concept Exercises on that track spending no more than 5-10 minutes solving each one.

I think this task might be stretching it a bit. Perhaps you could come up with a slightly simpler task?

p.s. we've recently started encouraging people to submit draft PR's with just three or four files:

- A `design.md` document
- An `instructions.md` document
- An example implementation
- An `introduction.md` document (optional)

The idea of these draft PR's is to get quick feedback on the direction an exercise is going in. This includes the complexity of the tasks, the validity of the story, the scope of the exercise, and such. It might be worth considering doing these draft PR's to lessen the chance of having to re-do work :)

ErikSchierboom replied:
@wneumann Thoughts on this?

wneumann replied:
Yeah. I really like the problem, but it is harder to explain than it is to actually solve, so it's probably better for a practice exercise. I'll work on a new exercise.

As for the draft PR, the main problem is that the instructions and example implementation are the last things I write.but I can try reversing that for future concept exercises.

ErikSchierboom replied:
> Yeah. I really like the problem, but it is harder to explain than it is to actually solve, so it's probably better for a practice exercise. I'll work on a new exercise.

Brilliant. Make sure to keep this version of the exercise somewhere safe to later be able to convert it to a practice exercise.

> As for the draft PR, the main problem is that the instructions and example implementation are the last things I write.but I can try reversing that for future concept exercises.

What is your usual order creating files for a Concept Exercise? The draft suggestion is just that: a suggestion. I mostly suggested this as I can feel it is not ideal when you have to change an exercise you've already invested quite some time in. A draft could possibly allow us to more quickly give feedback.

wneumann replied:
> What is your usual order creating files for a Concept Exercise? 

I usually start with the design to get the points to cover down, then I do the after to write out everything that needs to be covered. Then I edit it down for the introduction to remove a bunch of the details that aren't needed. Form there I do the introduction, the example and tests, and then the hints. But if there is an existing story for me to poach, then I can do the exercise first -- it's just that coming up with a good, non-contrived exercise with a good story that hits all the important points is tough.

Though I'm learning that a part of that is that I'm trying to pack too much into one exercise. I need to break some into smaller chunks and separate them into multiple concept exercises.

ErikSchierboom replied:
> it's just that coming up with a good, non-contrived exercise with a good story that hits all the important points is tough.

Yeah, this is definitely hard. Our [list of stories](https://github.com/exercism/v3/blob/master/reference/stories/README.md#stories) can possibly help here.

> Though I'm learning that a part of that is that I'm trying to pack too much into one exercise. I need to break some into smaller chunks and separate them into multiple concept exercises.

👍 This is definitely something that is different in v3, where exercises are quite small and very focused.

wneumann commented on languages/swift/exercises/concept/loops/.meta/design.md:
```diff
@@ -0,0 +1,25 @@
+
+- Loop-like constructs like map, reduce, filter, etc
+
+## Concepts
+
+- `loops`
+- `control-transfer`
+
+## Prerequisites
+
+- `conditionals`
+- `arrays`
+- `strings`
+- `tuples`
```
Added, including `ranges` requirement.

wneumann commented on languages/swift/exercises/concept/loops/.meta/design.md:
```diff
@@ -0,0 +1,25 @@
+## Out of scope
+
+- Loop-like constructs like map, reduce, filter, etc
+
```
Added.

wneumann commented:
@ErikSchierboom I replaced the old Santa/Hashing story/exercise with a new one that needs a review.

ErikSchierboom commented on languages/swift/exercises/concept/loops/Sources/loops/loops.swift:
```diff
@@ -0,0 +1,17 @@
+}
+
+func orderTracker(orders: [(drink: String, time: String)]) -> (
+  beer: (first: String, last: String, total: Int)?, soda: (first: String, last: String, total: Int)?
```
Is the `?` for an optional value? If so, should the `optionals` concept also be a prerequisite?

wneumann replied:
Initially `strings` (which was converted into `string-components`) was a prerequisite and that had `optionals` as a prerequisite so I left it out. But the new exercise doesn't use `string-components` anymore, so I added `optionals` back in.

ErikSchierboom commented on languages/swift/exercises/concept/loops/.docs/instructions.md:
````diff
@@ -0,0 +1,48 @@
+Implement the function `timeToPrepare(drinks: [String]) -> Double` which takes a non-empty array of drinks and returns the total time required to make the order.
+
+```swift
+timeToPrepare(drinks: ["beer", "frozen drink", "shot])
````
```suggestion
timeToPrepare(drinks: ["beer", "frozen drink", "shot"])
```

ErikSchierboom reviewed (approved):
This is an excellent exercise! I also really like the story, well done.

wneumann commented on languages/swift/exercises/concept/loops/.docs/after.md:
````diff
@@ -29,6 +29,8 @@ print("Liftoff!")
 ```
 
+# Add in bit about while let x = blah? { … }
+
````
I added the section in to the after in the latest commit.
